index_impl.go

```diff
@@ -460,16 +498,31 @@ func (i *indexImpl) SearchInContext(ctx context.Context, req *SearchRequest) (sr
 		coll = collector.NewTopNCollector(req.Size, req.From, req.Sort)
 	}
 
-	// open a reader for this search
-	indexReader, err := i.i.Reader()
-	if err != nil {
-		return nil, fmt.Errorf("error opening index reader %v", err)
+	var knnHits []*search.DocumentMatch
+	var ok bool
+	var skipKnnCollector bool
+	if req.PreSearchData != nil {
+		for k, v := range req.PreSearchData {
+			switch k {
+			case search.KnnPreSearchDataKey:
+				if v != nil {
+					knnHits, ok = v.([]*search.DocumentMatch)
+					if !ok {
+						return nil, fmt.Errorf("knn preSearchData must be of type []*search.DocumentMatch")
+					}
+				}
+				skipKnnCollector = true
```

Comment:
SearchInContext is run for all queries.
So shouldn't skipKNNCollector be true only if the KNNPreSearchKey is true and the user is allowed to run KNN queries?
If the latter is a valid condition, then there ought to be separate files which determine if skipKNNCollector should be true/false --> should always. be false for file with the !vectors tag.

Does this make sense?

Reply:
no like the idea is in the incoming presearchdata in the request i can give a null value for KnnPreSearchDataKey which means that the index (in the alias) did not provide any valid knn hits. So it must just do basic tfidf and skip the knn collector search.

Reply:
with the !vectors tag, the collector will never run because requestHasKNN() will always return false